utils/merge_test.go

```diff
@@ -0,0 +1,581 @@
+package utils
+
+import (
+	"github.com/stretchr/testify/require"
+	"testing"
+
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMergeWithStructs(t *testing.T) {
+	t.Run("merge identical structs", func(t *testing.T) {
+		setupStructs(t)
+
+		merged, err := mergeTestStructs(base, patch)
+		require.NoError(t, err)
+
+		assert.Equal(t, expectedMerged, *merged)
+	})
+
+	t.Run("merge identical structs as pointers", func(t *testing.T) {
+		setupStructs(t)
+
+		merged, err := mergeTestStructsPtrs(&base, &patch)
+		require.NoError(t, err)
+
+		assert.Equal(t, expectedMerged, *merged)
+	})
+
+	t.Run("different base vals are overwritten by patch - simple", func(t *testing.T) {
+		s1 := simple{42, NewBool(true), simple2{30}}
+		s2 := simple{13, NewBool(false), simple2{30}}
+		expected := simple{13, NewBool(false), simple2{30}}
+
+		merged, err := mergeSimple(s1, s2)
+		require.NoError(t, err)
+
+		assert.NotEqual(t, s1, *merged)
+		assert.Equal(t, expected, *merged)
+	})
+
+	t.Run("different base vals are overwritten by patch", func(t *testing.T) {
+		setupStructs(t)
+
+		base.F = 1342.12
+		base.Struct1.Pi = NewInt(937)
+		base.Struct1p.Ui = 734
+		base.Struct1.Struct2.Sli = []int{123123, 1243123}
+
+		merged, err := mergeTestStructs(base, patch)
+		require.NoError(t, err)
+
+		assert.NotEqual(t, base, *merged)
+		assert.Equal(t, patch, *merged)
+	})
+
+	t.Run("nil values in patch are ignored", func(t *testing.T) {
+		setupStructs(t)
+
+		patch.Pi = nil
+		patch.Struct1.Pi16 = nil
+
+		merged, err := mergeTestStructs(base, patch)
+		require.NoError(t, err)
+
+		assert.NotEqual(t, patch, *merged)
+		assert.Equal(t, expectedMerged, *merged)
+	})
+
+	t.Run("nil structs in patch are ignored", func(t *testing.T) {
+		setupStructs(t)
+
+		patch.Struct1p = nil
+		patch.Struct1.Struct2p = nil
+
+		merged, err := mergeTestStructs(base, patch)
+		require.NoError(t, err)
+
+		assert.NotEqual(t, patch, *merged)
+		assert.Equal(t, expectedMerged, *merged)
+	})
+
+	t.Run("nil slices in patch are ignored", func(t *testing.T) {
+		setupStructs(t)
+
+		patch.Sls = nil
+		patch.Struct1.Sli = nil
+		patch.Struct1.Struct2p.Slf = nil
+
+		merged, err := mergeTestStructs(base, patch)
+		require.NoError(t, err)
+
+		assert.NotEqual(t, patch, *merged)
+		assert.Equal(t, expectedMerged, *merged)
+	})
+
+	t.Run("nil maps in patch are ignored", func(t *testing.T) {
+		setupStructs(t)
+
+		patch.Msi = nil
+		patch.Mspi = nil
+		patch.Struct1.Mis = nil
+		patch.Struct1.Struct2p.Mspi = nil
+
+		merged, err := mergeTestStructs(base, patch)
+		require.NoError(t, err)
+
+		assert.NotEqual(t, patch, *merged)
+		assert.Equal(t, expectedMerged, *merged)
+	})
+}
+
+func setupConfig(t *testing.T) *model.Config {
+	t.Helper()
+
+	cfg := &model.Config{}
+	cfg.SetDefaults()
+	return cfg
+}
+
+func TestMergeWithConfigs(t *testing.T) {
```

Comment:
Can we move these tests into the `config` package instead? (I like that we're explicitly testing the merging of configs, in addition to the general case!)